File: web/src/render/TextLayout.ts
// TextLayout TS 版，镜像 plugin/src/main/java/moe/hikari/canvas/render/TextLayout.java
// 契约 docs/rendering.md §3。横排与竖排前后端一起做。

import type { TextElement } from '@/types/protocol';
import { advance } from './GlyphMetricsLut';

/** 单字 glyph 在画布坐标下的绘制点。x + baselineY 与后端 Java 含义一致。 */
export interface PositionedGlyph {
    ch: string;
    x: number;
    baselineY: number;
    /** 竖排用：true 表示该字符应绕自己的绘制点旋转 90°（CJK 全角标点）。 */
    rotated?: boolean;
    /**
     * 该 glyph 在 {@link TextElement#text} 中的原 char index（输入字符串）。
     * <p>仅前端字段——双端一致性不受影响（后端 Java {@code TextLayout.PositionedGlyph} 不带；
     * 仅 PreviewRenderer placeholder hint 用 srcIndex 与 interpolator segments 做 char range
     * 映射）。emoji surrogate pair / combining mark 暂按单 char index 处理。</p>
     */
    srcIndex?: number;
}

/** 跨字体统一的 ascent 比例（rendering.md §3.2）。 */
export const ASCENT_RATIO = 0.8;

/**
 * 前后端一致的"标准"字符宽度函数。
 *
 * <p>问题背景：浏览器 {@code ctx.measureText} 和 Java {@code FontMetrics.charWidth}
 * 即使加载同一 TTF/OTF 也会返回不同值（hinting / subpixel 策略不同）。layout
 * 的 softWrap 依赖 char width，差 1 px 就可能让前后端换行点不同。</p>
 *
 * <p>方案：两端不再读 font metrics，改用规则型宽度：</p>
 * <ul>
 *   <li>码点 {@code < 0x2E80}（ASCII + 拉丁 + 一般标点）：{@code round(fontSize * 0.5)}</li>
 *   <li>其他（CJK / Hiragana / Katakana / 全角）：{@code fontSize}</li>
 * </ul>
 *
 * <p>代价：非等宽字体（如思源黑）的视觉字间距可能略宽，但换行一致性优先。
 * 像素字体 Ark Pixel 本身 monospaced，canonical ≈ actual，无副作用。</p>
 */
export function canonicalCharWidth(ch: string, fontSize: number): number {
    const cp = ch.charCodeAt(0);
    if (cp < 0x2E80) return Math.round(fontSize * 0.5);
    return fontSize;
}

/**
 * 优先查 per-font GlyphMetricsLut（与后端 FontMetricsTable mirror），
 * 表缺 / 字符缺则 fallback 到 {@link canonicalCharWidth}。fontId 沿调用链从 TextElement 透传。
 */
export function charAdvance(fontId: string, ch: string, fontSize: number): number {
    const real = advance(fontId, ch, fontSize);
    if (real > 0) return real;
    return canonicalCharWidth(ch, fontSize);
}

/**
 * 行首禁则：半全角标点不许出现在行首，回溯到上一行末。
 * 去重（原把 11 个全角标点重复写两遍，indexOf 判定对重复不敏感，仅冗余）。
 * 保留一组全角 + 一组半角。顺序：） 】 」 』 。 ， 、 ？ ！ ： ；。
 * 与后端 TextLayout.java 的 LINE_START_FORBIDDEN 逐字节一致（双端镜像硬约束）。
 */
const LINE_START_FORBIDDEN = '）】」』。，、？！：；)].,!?:;';

export function layoutText(t: TextElement): PositionedGlyph[] {
    if (!t.text) return [];
    if (t.vertical) return layoutVertical(t);
    return layoutHorizontal(t);
}

// ---------- 横排 ----------

function layoutHorizontal(t: TextElement): PositionedGlyph[] {
    const fontSize = t.fontSize;
    const fontId = t.fontId;
    const boxW = t.w;
    const letterSpacing = t.letterSpacing;
    const lineHeightMul = t.lineHeight <= 0 ? 1.2 : t.lineHeight;
    const ascentPx = Math.round(fontSize * ASCENT_RATIO);
    const lineHeightPx = Math.max(1, Math.round(fontSize * lineHeightMul));

    // 1) 硬换行（保留 srcIndex 起点，便于 step 2 / 4 透传到 PositionedGlyph.srcIndex）
    type LineWithIndex = { text: string; srcStart: number };
    const paragraphs: LineWithIndex[] = [];
    {
        let cursor = 0;
        const parts = t.text.split('\n');
        for (let pi = 0; pi < parts.length; pi++) {
            paragraphs.push({ text: parts[pi], srcStart: cursor });
            cursor += parts[pi].length + 1; // +1 for \n
        }
    }

    // 2) 软换行
    const lines: LineWithIndex[] = [];
    for (const para of paragraphs) {
        if (para.text === '') {
            lines.push({ text: '', srcStart: para.srcStart });
        } else {
            softWrap(para.text, para.srcStart, fontId, fontSize, boxW, letterSpacing, lines);
        }
    }

    // 3) 行首禁则
    applyLineStartForbidden(lines);

    // 4) align + 基线定位 + letterSpacing 逐字符累加（同步携带 srcIndex）
    const out: PositionedGlyph[] = [];
    for (let li = 0; li < lines.length; li++) {
        const line = lines[li];
        const lineTopY = t.y + li * lineHeightPx;
        const baselineY = lineTopY + ascentPx;
        const lineWidthPx = measureLineWidth(line.text, fontId, fontSize, letterSpacing);
        let startX = t.x;
        // 用 Math.trunc（向零截断）匹配后端 Java 整数除法 `/2` 语义，
        // 避免行宽溢出文本框（负奇数分子）时双端 startX 差 1px。
        if (t.align === 'center') startX = t.x + Math.trunc((boxW - lineWidthPx) / 2);
        else if (t.align === 'right') startX = t.x + boxW - lineWidthPx;

        let cursorX = startX;
        for (let i = 0; i < line.text.length; i++) {
            const ch = line.text[i];
            out.push({ ch, x: cursorX, baselineY, srcIndex: line.srcStart + i });
            cursorX += charAdvance(fontId, ch, fontSize);
            if (i < line.text.length - 1) cursorX += Math.round(letterSpacing);
        }
    }
    return out;
}

function softWrap(text: string, srcStart: number, fontId: string, fontSize: number, maxW: number,
                  letterSpacing: number, out: { text: string; srcStart: number }[]): void {
    const n = text.length;
    let cursor = 0;
    while (cursor < n) {
        let accW = 0;
        let lastBreakEnd = -1;
        let breakAtWhitespace = false;
        let i = cursor;
        let breakOut = -1;

        while (i < n) {
            const ch = text[i];
            const cw = charAdvance(fontId, ch, fontSize);
            const step = cw + (i > cursor ? Math.round(letterSpacing) : 0);
            if (accW + step > maxW && i > cursor) {
                breakOut = i;
                break;
            }
            accW += step;
            if (ch === ' ' || ch === '\u3000') {
                lastBreakEnd = i + 1;
                breakAtWhitespace = true;
            } else if (isCjk(ch)) {
                lastBreakEnd = i + 1;
                breakAtWhitespace = false;
            }
            i++;
        }

        if (breakOut < 0) {
            out.push({ text: text.substring(cursor), srcStart: srcStart + cursor });
            return;
        }

        let nextStart: number;
        if (lastBreakEnd >= cursor && lastBreakEnd <= breakOut) {
            const endOfLine = breakAtWhitespace ? lastBreakEnd - 1 : lastBreakEnd;
            out.push({ text: text.substring(cursor, endOfLine), srcStart: srcStart + cursor });
            nextStart = lastBreakEnd;
            while (nextStart < n && text[nextStart] === ' ') nextStart++;
        } else {
            out.push({ text: text.substring(cursor, breakOut), srcStart: srcStart + cursor });
            nextStart = breakOut;
        }
        cursor = nextStart;
    }
}

function applyLineStartForbidden(lines: { text: string; srcStart: number }[]): void {
    for (let i = 1; i < lines.length; i++) {
        const line = lines[i];
        if (line.text.length === 0) continue;
        const first = line.text[0];
        if (LINE_START_FORBIDDEN.indexOf(first) >= 0) {
            // \u628a\u9996\u5b57\u7b26\u56de\u632a\u5230\u4e0a\u4e00\u884c\u672b\u5c3e\u2014\u2014\u4e0a\u4e00\u884c\u4e0d\u52a8 srcStart\uff0c\u672c\u884c srcStart +1
            lines[i - 1] = { text: lines[i - 1].text + first, srcStart: lines[i - 1].srcStart };
            lines[i] = { text: line.text.substring(1), srcStart: line.srcStart + 1 };
        }
    }
}

function measureLineWidth(line: string, fontId: string, fontSize: number, letterSpacing: number): number {
    if (line.length === 0) return 0;
    let w = 0;
    for (let i = 0; i < line.length; i++) {
        w += charAdvance(fontId, line[i], fontSize);
        if (i < line.length - 1) w += Math.round(letterSpacing);
    }
    return w;
}

// ---------- 竖排（rendering.md §3.3，与 Java TextLayout.layoutVertical 镜像） ----------

function layoutVertical(t: TextElement): PositionedGlyph[] {
    const fontSize = t.fontSize;
    const fontId = t.fontId;
    const letterSpacing = Math.round(t.letterSpacing);
    const lineHeightMul = t.lineHeight <= 0 ? 1.2 : t.lineHeight;
    const colStep = Math.max(1, Math.round(fontSize * lineHeightMul));
    const ascentPx = Math.round(fontSize * ASCENT_RATIO);
    const boxH = t.h;

    // 同 horizontal：携带 srcStart（每个 col 的起始 char 在原 t.text 中 index）
    type Col = { chars: string[]; srcStart: number };
    const columns: Col[] = [];
    {
        let cursor = 0;
        const parts = t.text.split('\n');
        for (let pi = 0; pi < parts.length; pi++) {
            const para = parts[pi];
            if (para === '') {
                columns.push({ chars: [], srcStart: cursor });
            } else {
                softWrapVertical(para, cursor, fontSize, letterSpacing, boxH, columns);
            }
            cursor += para.length + 1;
        }
    }

    const out: PositionedGlyph[] = [];
    const totalCols = columns.length;
    for (let ci = 0; ci < totalCols; ci++) {
        const col = columns[ci];
        if (col.chars.length === 0) continue;
        const colCenterX = t.x + t.w - ci * colStep - Math.floor(colStep / 2);
        const cellsH = col.chars.length;
        const totalH = cellsH * fontSize + Math.max(0, cellsH - 1) * letterSpacing;
        let startTopY: number;
        // 同横排——竖排 center 起点用 Math.trunc 匹配后端 `/2` 截断语义。
        if (t.align === 'center') startTopY = t.y + Math.trunc((boxH - totalH) / 2);
        else if (t.align === 'right') startTopY = t.y + boxH - totalH;
        else startTopY = t.y;

        let cellTopY = startTopY;
        for (let i = 0; i < col.chars.length; i++) {
            const ch = col.chars[i];
            const srcIndex = col.srcStart + i;
            if (isRotatableVertical(ch)) {
                out.push({ ch, x: colCenterX, baselineY: cellTopY + Math.floor(fontSize / 2), rotated: true, srcIndex });
            } else {
                const chW = charAdvance(fontId, ch, fontSize);
                out.push({ ch, x: colCenterX - Math.floor(chW / 2), baselineY: cellTopY + ascentPx, srcIndex });
            }
            cellTopY += fontSize + letterSpacing;
        }
    }
    return out;
}

function softWrapVertical(text: string, srcStart: number, fontSize: number, letterSpacing: number,
                          maxH: number, cols: { chars: string[]; srcStart: number }[]): void {
    const n = text.length;
    let i = 0;
    while (i < n) {
        const chars: string[] = [];
        const colStartChar = i;
        let accH = 0;
        while (i < n) {
            const step = fontSize + (chars.length === 0 ? 0 : letterSpacing);
            if (accH + step > maxH && chars.length > 0) break;
            chars.push(text[i]);
            accH += step;
            i++;
        }
        cols.push({ chars, srcStart: srcStart + colStartChar });
    }
}

/** 竖排下应旋转 90° 的字符：全角标点与半宽全宽形式（与 Java isRotatableVertical 一致）。 */
function isRotatableVertical(ch: string): boolean {
    const cp = ch.charCodeAt(0);
    return (cp >= 0x3000 && cp <= 0x303F) || (cp >= 0xFF00 && cp <= 0xFFEF);
}

// ---------- CJK 判断（与 Java 版相同范围）----------

function isCjk(c: string): boolean {
    const cp = c.charCodeAt(0);
    return (cp >= 0x4E00 && cp <= 0x9FFF)   // CJK Unified
        || (cp >= 0x3400 && cp <= 0x4DBF) // Ext A
        || (cp >= 0x3040 && cp <= 0x309F) // Hiragana
        || (cp >= 0x30A0 && cp <= 0x30FF) // Katakana
        || (cp >= 0xFF00 && cp <= 0xFFEF) // Halfwidth/Fullwidth
        || (cp >= 0x3000 && cp <= 0x303F); // CJK Punct
}

// ---------- CanvasMeasurer 已废弃（切到 canonicalCharWidth） ----------
// 原文件导出的 CanvasMeasurer 类与 CharMeasurer 接口此轮移除，调用方已不再传 measurer。

File: plugin/src/main/java/moe/hikari/canvas/render/TextLayout.java
package moe.hikari.canvas.render;

import moe.hikari.canvas.state.TextElement;

import java.awt.FontMetrics;
import java.util.ArrayList;
import java.util.List;

/**
 * 横排文本排版算法，契约对应 {@code docs/rendering.md §3}。
 *
 * <h2>步骤</h2>
 * <ol>
 *   <li>按 {@code \n} 切硬换行段</li>
 *   <li>按 {@link TextElement#w()} 做软换行（CJK 任意位置、空白字符前、拉丁文空白 break）</li>
 *   <li>应用行首禁则：{@code ）】」』。，、？！：；} 等标点不允许独占行首，回溯到上一行末尾</li>
 *   <li>对每行应用 {@link TextElement#align()} 对齐</li>
 *   <li>计算基线：每行高 = {@code fontSize × lineHeight}；基线 = 行顶 + {@code fontSize × ascentRatio}（rendering.md §3.2 固定 0.8）</li>
 * </ol>
 *
 * <h2>不处理</h2>
 * 竖排（{@code vertical=true}）：M4-T5 不实装，由 {@link CanvasCompositor} 在调用前检测并 WARN；
 * 真竖排排版推迟到 M4.5 / M7，见 {@code docs/rendering.md §3.3}。
 *
 * <h2>纯函数</h2>
 * 静态方法，无状态，线程安全。
 */
public final class TextLayout {

    /** 跨字体统一的 ascent 比例，见 rendering.md §3.2。 */
    public static final double ASCENT_RATIO = 0.8;

    /** 行首禁则：出现在行首的标点会被回溯到上一行末尾。半全角都收录。 */
    private static final String LINE_START_FORBIDDEN = "）】」』。，、？！：；）】」』。，、？！：；)].,!?:;";

    private TextLayout() {}

    /**
     * 单个字符 glyph 在画布坐标系下的绘制位置。
     *
     * <p>语义按 {@code rotated}：</p>
     * <ul>
     *   <li>{@code rotated == false}：{@code (x, baselineY)} 是 Graphics2D.drawString 的标准
     *       锚点——字符 baseline 起点；调用方直接 {@code g.drawString(ch, x, baselineY)}</li>
     *   <li>{@code rotated == true}（M5-C3 竖排全角标点）：{@code (x, baselineY)} 是字符
     *       <b>方格中心</b>；调用方应 translate 到该点 → rotate(π/2) → drawString 在字符
     *       自身坐标系中心偏移处（参见 CanvasCompositor.drawRotatedGlyph）</li>
     * </ul>
     */
    public record PositionedGlyph(String ch, int x, int baselineY, boolean rotated) {
        public PositionedGlyph(String ch, int x, int baselineY) {
            this(ch, x, baselineY, false);
        }
    }

    /**
     * 给定 TextElement + FontMetrics（来自目标 fontSize 下派生的 Font），
     * 返回每个字符的绘制位置列表。{@code letterSpacing} 已累计进 x。
     *
     * <p>调用方用 {@code Graphics2D.drawString(glyph.ch(), glyph.x(), glyph.baselineY())}
     * 逐字符绘制。之所以不直接返回整行 + 一次 drawString，是为了支持 letterSpacing
     * 对整数像素的独立控制（Graphics2D 本身不支持 per-char letterSpacing）。</p>
     */
    public static List<PositionedGlyph> layout(TextElement t, FontMetrics fm) {
        if (t.text() == null || t.text().isEmpty()) {
            return List.of();
        }
        if (t.vertical()) {
            return layoutVertical(t, fm);
        }
        int fontSize = t.fontSize();
        int boxWidth = t.w();
        float letterSpacing = t.letterSpacing();
        float lineHeightMul = t.lineHeight() <= 0 ? 1.2f : t.lineHeight();
        int ascentPx = (int) Math.round(fontSize * ASCENT_RATIO);
        int lineHeightPx = Math.max(1, Math.round(fontSize * lineHeightMul));

        // 1) 硬换行
        String[] paragraphs = t.text().split("\n", -1);

        // 2) 软换行（逐段）
        List<String> lines = new ArrayList<>();
        for (String para : paragraphs) {
            if (para.isEmpty()) {
                lines.add("");
            } else {
                softWrap(para, fm, boxWidth, letterSpacing, lines);
            }
        }

        // 3) 行首禁则：回溯下一行首禁止标点到上一行末
        applyLineStartForbidden(lines);

        // 4) 按对齐 + 基线定位
        List<PositionedGlyph> out = new ArrayList<>(t.text().length());
        for (int li = 0; li < lines.size(); li++) {
            String line = lines.get(li);
            int lineTopY = t.y() + li * lineHeightPx;
            int baselineY = lineTopY + ascentPx;
            int lineWidthPx = measureLineWidth(line, fm, letterSpacing);
            int startX = switch (t.align()) {
                case "center" -> t.x() + (boxWidth - lineWidthPx) / 2;
                case "right" -> t.x() + boxWidth - lineWidthPx;
                default -> t.x(); // "left" + 未知值 fallback
            };
            int cursorX = startX;
            for (int i = 0; i < line.length(); i++) {
                char c = line.charAt(i);
                out.add(new PositionedGlyph(String.valueOf(c), cursorX, baselineY));
                cursorX += fm.charWidth(c);
                if (i < line.length() - 1) {
                    cursorX += Math.round(letterSpacing);
                }
            }
        }
        return out;
    }

    /**
     * 软换行：在 {@code maxWidth} 内逐字符累计宽度，触发换行时回溯到最近可断点。
     *
     * <p>断点规则：</p>
     * <ul>
     *   <li>{@code ' '}（半角空格）/ {@code '\u3000'}（全角空格）：其 <b>后</b> 可断</li>
     *   <li>CJK 字符：其 <b>后</b> 随时可断（任意两字之间）</li>
     *   <li>若整段无可断点（全 ASCII 长词），强制在当前位置截断</li>
     * </ul>
     *
     * <p>换行时：若断点为空白字符，空白字符 <b>丢弃</b>（不出现在下一行行首）。</p>
     */
    private static void softWrap(String text, FontMetrics fm, int maxWidth,
                                 float letterSpacing, List<String> out) {
        int n = text.length();
        int cursor = 0;
        while (cursor < n) {
            int accWidth = 0;
            int lastBreakEnd = -1; // 最近可断点之后的位置（包含断点字符本身，下一行从这里开始）
            boolean breakAtWhitespace = false;
            int i = cursor;
            int breakOut = -1;

            while (i < n) {
                char c = text.charAt(i);
                int cw = fm.charWidth(c);
                int step = cw + (i > cursor ? Math.round(letterSpacing) : 0);
                if (accWidth + step > maxWidth && i > cursor) {
                    breakOut = i;
                    break;
                }
                accWidth += step;
                if (c == ' ' || c == '\u3000') {
                    lastBreakEnd = i + 1;
                    breakAtWhitespace = true;
                } else if (isCjk(c)) {
                    lastBreakEnd = i + 1;
                    breakAtWhitespace = false;
                }
                i++;
            }

            if (breakOut < 0) {
                // 整段装下
                out.add(text.substring(cursor));
                return;
            }

            int nextStart;
            if (lastBreakEnd >= cursor && lastBreakEnd <= breakOut) {
                // 有可断点
                int endOfLine = breakAtWhitespace ? lastBreakEnd - 1 : lastBreakEnd; // 末尾去空白
                out.add(text.substring(cursor, endOfLine));
                nextStart = lastBreakEnd;
                // 连续空白全部吞掉（避免下一行以空白开头）
                while (nextStart < n && text.charAt(nextStart) == ' ') {
                    nextStart++;
                }
            } else {
                // 无断点，硬截断
                out.add(text.substring(cursor, breakOut));
                nextStart = breakOut;
            }
            cursor = nextStart;
        }
    }

    /** 行首禁则：若下一行首字符属禁止集合，把它挪到上一行末尾。 */
    private static void applyLineStartForbidden(List<String> lines) {
        for (int i = 1; i < lines.size(); i++) {
            String line = lines.get(i);
            if (line.isEmpty()) continue;
            char first = line.charAt(0);
            if (LINE_START_FORBIDDEN.indexOf(first) >= 0) {
                lines.set(i - 1, lines.get(i - 1) + first);
                lines.set(i, line.substring(1));
            }
        }
    }

    private static int measureLineWidth(String line, FontMetrics fm, float letterSpacing) {
        if (line.isEmpty()) return 0;
        int width = 0;
        for (int i = 0; i < line.length(); i++) {
            width += fm.charWidth(line.charAt(i));
            if (i < line.length() - 1) {
                width += Math.round(letterSpacing);
            }
        }
        return width;
    }

    /**
     * 粗略判断 CJK 统一汉字 + 日文假名 + 全角符号范围。
     * 够用级判断，不覆盖所有 Unicode CJK 扩展。
     */
    private static boolean isCjk(char c) {
        return (c >= '\u4E00' && c <= '\u9FFF')   // CJK Unified Ideographs
                || (c >= '\u3400' && c <= '\u4DBF') // CJK Extension A
                || (c >= '\u3040' && c <= '\u309F') // Hiragana
                || (c >= '\u30A0' && c <= '\u30FF') // Katakana
                || (c >= '\uFF00' && c <= '\uFFEF') // Halfwidth/Fullwidth Forms
                || (c >= '\u3000' && c <= '\u303F'); // CJK Symbols & Punctuation
    }

    // ---------- M5-C6 竖排（rendering.md §3.3） ----------

    /**
     * 竖排布局。与横排不同：
     * <ul>
     *   <li>字符从上到下、列从右到左（CJK 传统）</li>
     *   <li>每个字符占 {@code fontSize × fontSize} 方格；列宽 = {@code fontSize * lineHeight}</li>
     *   <li>全角标点（{@code U+3000-U+303F} / {@code U+FF00-U+FFEF}）旋转 90° → {@code PositionedGlyph.rotated = true}</li>
     *   <li>半角字符不旋转（直立）</li>
     *   <li>{@code align} 在竖排下语义 = 列内 <b>顶/中/底</b> 对齐</li>
     *   <li>软换行：按 box {@code h}；硬换行 {@code \n} 起新列</li>
     * </ul>
     *
     * <p>行首禁则在竖排下未实装（M7 polish）——相对少见。</p>
     */
    private static List<PositionedGlyph> layoutVertical(TextElement t, FontMetrics fm) {
        int fontSize = t.fontSize();
        int letterSpacing = Math.round(t.letterSpacing());
        float lineHeightMul = t.lineHeight() <= 0 ? 1.2f : t.lineHeight();
        int colStep = Math.max(1, Math.round(fontSize * lineHeightMul));
        int ascentPx = (int) Math.round(fontSize * ASCENT_RATIO);
        int boxH = t.h();

        String[] paragraphs = t.text().split("\n", -1);
        List<List<String>> columns = new ArrayList<>();
        for (String para : paragraphs) {
            if (para.isEmpty()) {
                columns.add(List.of());
            } else {
                softWrapVertical(para, fontSize, letterSpacing, boxH, columns);
            }
        }

        List<PositionedGlyph> out = new ArrayList<>(t.text().length());
        int totalCols = columns.size();
        for (int ci = 0; ci < totalCols; ci++) {
            List<String> col = columns.get(ci);
            if (col.isEmpty()) continue;
            int colCenterX = t.x() + t.w() - ci * colStep - colStep / 2;
            int cellsH = col.size();
            int totalH = cellsH * fontSize + Math.max(0, cellsH - 1) * letterSpacing;
            int startTopY;
            if ("center".equals(t.align())) startTopY = t.y() + (boxH - totalH) / 2;
            else if ("right".equals(t.align())) startTopY = t.y() + boxH - totalH;
            else startTopY = t.y();

            int cellTopY = startTopY;
            for (String s : col) {
                char c = s.charAt(0);
                if (isRotatableVertical(c)) {
                    // pivot = 方格中心；CanvasCompositor 绕此点 rotate 90° 后 drawString
                    out.add(new PositionedGlyph(s, colCenterX, cellTopY + fontSize / 2, true));
                } else {
                    int chW = fm.charWidth(c);
                    out.add(new PositionedGlyph(s, colCenterX - chW / 2, cellTopY + ascentPx, false));
                }
                cellTopY += fontSize + letterSpacing;
            }
        }
        return out;
    }

    private static void softWrapVertical(String text, int fontSize, int letterSpacing,
                                         int maxH, List<List<String>> cols) {
        int n = text.length();
        int i = 0;
        while (i < n) {
            List<String> col = new ArrayList<>();
            int accH = 0;
            while (i < n) {
                int step = fontSize + (col.isEmpty() ? 0 : letterSpacing);
                if (accH + step > maxH && !col.isEmpty()) break;
                col.add(String.valueOf(text.charAt(i)));
                accH += step;
                i++;
            }
            cols.add(col);
        }
    }

    /**
     * 判断字符是否应在竖排下旋转 90°。
     * 覆盖 CJK 标点与符号（U+3000-U+303F）+ 半宽全宽形式（U+FF00-U+FFEF，含全角括号等）。
     * 全角汉字（U+4E00+）本身方形，不旋转。
     */
    private static boolean isRotatableVertical(char c) {
        return (c >= '\u3000' && c <= '\u303F')
                || (c >= '\uFF00' && c <= '\uFFEF');
    }
}
